pkg/kds/context/context.go

```diff
@@ -89,59 +89,61 @@ func MapZoneTokenSigningKeyGlobalToPublicKey(
 }
 
 // GlobalProvidedFilter returns ResourceFilter which filters Resources provided by Global, specifically
-// excludes Dataplanes and Ingresses from 'clusterID' cluster
+// excludes Dataplanes, Ingresses and Egresses from 'clusterID' cluster
 func GlobalProvidedFilter(rm manager.ResourceManager, configs map[string]bool) reconcile.ResourceFilter {
 	return func(clusterID string, r model.Resource) bool {
-		resType := r.Descriptor().Name
 		resName := r.GetMeta().GetName()
 
-		if resType == system.ConfigType && !configs[resName] {
+		switch r.Descriptor().Name {
+		case mesh.DataplaneType:
 			return false
-		}
-		if resType == system.GlobalSecretType {
+		case system.ConfigType:
+			return configs[resName]
+		case system.GlobalSecretType:
 			return util.ResourceNameHasAtLeastOneOfPrefixes(
 				resName,
 				zoneingress.ZoneIngressSigningKeyPrefix,
 				zone_tokens.SigningKeyPrefix,
 			)
-		}
-		if resType != mesh.DataplaneType && resType != mesh.ZoneIngressType {
-			return true
-		}
-		if resType == mesh.DataplaneType {
-			return false
-		}
-		if clusterID == util.ZoneTag(r) {
-			// don't need to sync resource to the zone where resource is originated from
-			return false
-		}
-		zone := system.NewZoneResource()
-		if err := rm.Get(context.Background(), zone, store.GetByKey(util.ZoneTag(r), model.NoMesh)); err != nil {
-			log.Error(err, "failed to get zone", "zone", util.ZoneTag(r))
-			// since there is no explicit 'enabled: false' then we don't
-			// make any strong decisions which might affect connectivity
+		case mesh.ZoneIngressType, mesh.ZoneEgressType:
+			zoneTag := util.ZoneTag(r)
+
+			if clusterID == zoneTag {
+				// don't need to sync resource to the zone where resource is originated from
+				return false
+			}
+
```

Comment:
Something is wrong here. I think the code above in this case should go even before switch.

It also seems like we don't have any test for this since we just changed the behavior to execute this logic only in case of ZoneIngress and Egress and nothing is broken.

Paging @lobkovilya for clarification.

Reply:
I've spent 2h actually trying to understand it and that's the result of it, which I also wrote tests for.
See that https://github.com/kumahq/kuma/pull/3757/files/bc5c5ac06c21ed79953f81993806f3edb969a68e#diff-f3aee24192920832abfb78a72f0f246eaa844b77862826920b59a159ac587d65L108 is filtering out any resources which are not dataplanes or zone ingresses (now zone egresses as well) and returns `true` for them, the dataplanes returns `false` according to code three lines below, so the only resources which are left for this function are zone ingresses (and zone egresses now), that's why I rewrote it to be more explicit

Reply:
Right, but what I'm questioning is the logic that we have in master. AFAIR the reason for introducing this zone#isEnabled was to not sync resources when zone is disabled. So why the logic is only executed on some of the resources

Reply:
@lobkovilya wdyt?

Reply:
If the zone is disabled then we delete ingresses of that zone from every other zone. We don't stop syncing policies to the disabled zone. We have a [test](https://github.com/kumahq/kuma/blob/master/test/e2e/deploy/kuma_deploy_universal.go#L159) that checks this behavior.

Reply:
I see, so so even if zone is disabled, we accept all resources from the zone to global, but we don't sync zone ingresses to other zones. I think we need to clarify this in docs.

@bartsmykla with that in mind, your code is almost correct. We just need to remove `mesh.ZoneEgressType` from this case, since we don't sync ZoneEgress from global to zone anyways.

Reply:
ok, addressed